jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/GraphOrdering.java

```diff
@@ -0,0 +1,269 @@
+    private Object[]        mapOrderToVertex;
+    private int             vertexCount;
+
+    private int[][]         outgoingEdges;
+    private int[][]         incomingEdges;
+    private Boolean[][]     adjMatrix;
+
+    private boolean         cacheEdges;
+
+
+    /**
+     * @param graph the graph to be ordered
+     * @param orderByDegree should the vertices be ordered by their degree. This
+     *        speeds up the VF2 algorithm.
+     * @param cacheEdges if true, the class creates a adjacency matrix and two
+     *        arrays for incoming and outgoing edges for fast access.
+     */
+    public GraphOrdering(Graph<V, E> graph,
+                         boolean orderByDegree,
+                         boolean cacheEdges) {
+        this.graph      = graph;
+        this.cacheEdges = cacheEdges;
+
+        List<V> vertexSet = new ArrayList<V>(graph.vertexSet());
+        if (orderByDegree)
+            java.util.Collections.sort(vertexSet,
+                            new GeneralVertexDegreeComparator<V>(graph));
+
+        vertexCount      = vertexSet.size();
+        mapVertexToOrder = new HashMap<V, Integer>();
+        mapOrderToVertex = new Object[vertexCount];
+
+        outgoingEdges    = new int[vertexCount][];
+        incomingEdges    = new int[vertexCount][];
+        adjMatrix        = new Boolean[vertexCount][vertexCount];
+
+        Integer i = 0;
+        for (V vertex : vertexSet) {
+            mapVertexToOrder.put(vertex, i);
+            mapOrderToVertex[i] = vertex;
+
+            outgoingEdges[i]   = null;
+            incomingEdges[i++] = null;
+        }
+    }
+
+
+    /**
+     * @param graph the graph to be ordered
+     */
+    public GraphOrdering(Graph<V, E> graph) {
+        this(graph, false, true);
+    }
+
+    /**
+     * @return returns the number of vertices in the graph.
+     */
+    public int getVertexCount() {
+        return this.vertexCount;
+    }
+
+    /**
+     * @param vertexNumber the number which identifies the vertex v in this
+     *        order.
+     * @return the identifying numbers of all vertices which are connected to v
+     *         by an edge outgoing from v.
+     */
+    public int[] getOutEdges(int vertexNumber) {
+        if (cacheEdges && outgoingEdges[vertexNumber] != null)
+            return outgoingEdges[vertexNumber];
+
+        V v = getVertex(vertexNumber);
+        Set<E> edgeSet = null;
+
+        if (graph instanceof DirectedGraph<?, ?>)
+            edgeSet = ((DirectedGraph<V, E>) graph).outgoingEdgesOf(v);
+        else
+            edgeSet = graph.edgesOf(v);
+
+        int[] vertexArray = new int[edgeSet.size()];
+        int i = 0;
+
+        for (E edge : edgeSet) {
+            V source = graph.getEdgeSource(edge),
+              target = graph.getEdgeTarget(edge);
+            vertexArray[i++] =
+                mapVertexToOrder.get(source.equals(v) ? target : source);
+        }
+
+        return outgoingEdges[vertexNumber] = vertexArray;
```

Comment:
It's a bit odd that cacheEdges only controls the read path (not the write path).